File: tests/core/test_dyncalls.js
mergeInto(LibraryManager.library, {
  test_dyncalls_vijdf: function(funcPtr) {
    // 1. Directly access a function pointer via a static signature (32-bit ABI)
    //    (this is the fastest way to call a function pointer when the signature is statically known in WASM_BIGINT==0 builds)
    dynCall_vijdf(funcPtr, 1, /*lo=*/2, /*hi=*/3, 4, 5); // Available only in WASM_BIGINT != 2 builds

    // 2. Access a function pointer using the convenience/legacy 'dynCall' function (32-bit ABI)
    //    (this form should never be used, it is suboptimal for performance, but provided for legacy compatibility)
    dynCall('vijdf', funcPtr, [2, /*lo=*/3, /*hi=*/4, 5, 6]); // Available only in WASM_BIGINT != 2 builds
  },

  test_dyncalls_iii: function(funcPtr) {
    // 1. Directly access a function pointer via a static signature (32-bit ABI)
    //    (this is the fastest way to call a function pointer when the signature is statically known in WASM_BIGINT==0 builds)
    var ret = dynCall_iii(funcPtr, 1, 2); // Available only in WASM_BIGINT != 2 builds
    console.log('iii returned ' + ret);

    // 2. Access a function pointer using the convenience/legacy 'dynCall' function (32-bit ABI)
    //    (this form should never be used, it is suboptimal for performance, but provided for legacy compatibility)
    var ret = dynCall('iii', funcPtr, [2, 3]); // Available only in WASM_BIGINT != 2 builds
    console.log('iii returned ' + ret);
  }
});

File: tests/core/test_dyncalls.c
#include <emscripten.h>
#include <stdint.h>
#include <stdio.h>

void vijdf(int i, int64_t j, double d, float f)
{
	EM_ASM(console.log('vijdf: i='+$0+',jlo='+$1+',jhi='+$2+',d='+$3+',f='+$4), i, (uint32_t)j, (uint32_t)(((uint64_t)j) >> 32), d, f);
}

int iii(int i, int j)
{
	EM_ASM(console.log('iii: i='+$0+',j='+$1), i, j);
	return 42;
}


void test_dyncalls_vijdf(void(*)(int, int64_t, double, float));
void test_dyncalls_iii(int(*)(int, int));

int main()
{
	test_dyncalls_vijdf(&vijdf);
	test_dyncalls_iii(&iii);
}
